### PR TITLE
Add CSV support

### PR DESCRIPTION
**Pull request by kevinkhao**
This PR does a few things:
- Mainly, add CSV support
- Refactor, clean some details in base module and in XLSX module for better separation
- Various misc. fixes and improvements (comments, minor details)

### CODE REVIEW

**Comment by kevinkhao** [comment]
Currently works with specifying two parameters: value delimiter and quotechar. Fields are invisible when format is not CSV. Default values are the simple Excel dialect ("," and " " ")

Screenshot from 2020-09-30 18-44-48

**Comment by kevinkhao** [comment]
https://github.com/akretion/pattern-import-export/pull/18/commits/068d44b1a2e9cbda2a07b3c61ab0dc10da0141f5 improves an error but it still technically could be a problem. To explain: it is a simple move of the rest of the function into the "try" clause.

It is possible to import a CSV in wrong format, and because there are no delimiters detected, it will treat the whole line as 1 value (because it did not detect the correct delimiters).

Example:
Delimiter is configured as ; but my .csv has only , as delimiter
id;country_id
4;23
The importer will see one field with name "id;country_id" and one line with one value "4;23"
The problem is not detected during data extraction, but during load(). So, it is very unlikely but still possible to do bad loads when delimiters are not configured correctly.

File: pattern_import_export/models/ir_exports.py [hunk 1]
```diff
@@ -232,19 +232,20 @@ def _generate_import_with_pattern_job(self, patterned_import):
         try:
             attachment_data = base64.b64decode(patterned_import.datas.decode("utf-8"))
             datas = self._read_import_data(attachment_data)
+            res = (
+                self.with_context(
+                    load_format="flatty",
```
Comment (sebastienbeau) [inline pattern_import_export/models/ir_exports.py]: can you rebase your PR and adapt the code (flatty key is now removed)

https://github.com/akretion/pattern-import-export/pull/17/files#diff-b0c82f1e07e5526647175a9398c7a32dR244

File: pattern_import_export_csv/models/ir_exports.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        file_fmtd = io.StringIO(bytes_content.decode("utf-8"))
+        return csv.DictReader(
+            file_fmtd,
+            delimiter=self.csv_value_delimiter,
+            quotechar=self.csv_quote_character,
+        )
+
+    # Export part
+
+    def _write_headers(self, writer):
```
Comment (sebastienbeau) [inline pattern_import_export_csv/models/ir_exports.py]: maybe _csv_write_headers ? (to avoid possible conflict naming)

Comment (sebastienbeau) [reply pattern_import_export_csv/models/ir_exports.py]: Or maybe the code used for generating should be move in a wizard?

File: pattern_import_export_csv/models/ir_exports.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        if self.use_description:
+            writer.writerow(self._get_header(use_description=True))
+        writer.writerow(self._get_header(False))
+
+    def _write_rows(self, writer, records):
+        for row in self._get_data_to_export(records):
+            writer.writerow(row.values())
```
Comment (sebastienbeau) [inline pattern_import_export_csv/models/ir_exports.py]: TODO check if we do not have issue with order here?
Maybe using dictwriter is a better solution ?

File: pattern_import_export_csv/models/ir_exports.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        return stringio.getvalue().encode("utf_8")
+
+    def _attachment_to_csv_writer(self, attachment):
+        # Convention for editing CSV file is to slurp it into memory then
+        #   modify it, then save it into a new file
+        bytes_content = base64.b64decode(attachment.datas)
+        return self._bytes_to_csv_writer(bytes_content)
+
+    def _bytes_to_csv_reader(self, bytes_content):
+        file_fmtd = io.StringIO(bytes_content.decode("utf-8"))
```
Comment (sebastienbeau) [inline pattern_import_export_csv/models/ir_exports.py:44]: use dictreader

File: pattern_import_export_csv/models/ir_exports.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            delimiter=self.csv_value_delimiter,
+            quotechar=self.csv_quote_character,
+        )
+        self._write_headers(writer)
+        self._write_rows(writer, records)
+        return self._csv_stringio_to_bytes(virtual_file)
+
+    # Import part
+
+    def _check_csv_has_error_column(self, attachment):
```
Comment (sebastienbeau) [inline pattern_import_export_csv/models/ir_exports.py]: let's remove the error column

File: pattern_import_export/models/ir_exports.py (hunk 1 above)
Comment (kevinkhao) [reply pattern_import_export/models/ir_exports.py]: done

File: pattern_import_export_csv/models/ir_exports.py (hunk 2 above)
Comment (kevinkhao) [reply pattern_import_export_csv/models/ir_exports.py]: done

File: pattern_import_export_csv/models/ir_exports.py (hunk 5 above)
Comment (kevinkhao) [reply pattern_import_export_csv/models/ir_exports.py]: done

**Comment by sebastienbeau** [comment]
@kevinkhao can you do a rebase on shopinvader/12.0 and do a PR on shopinvader directly?

**Comment by kevinkhao** [comment]
Done, https://github.com/shopinvader/pattern-import-export/pull/32

**Pull request closed without merging**